lib/Sema/TypeCheckDecl.cpp

```diff
@@ -242,31 +240,31 @@ void TypeChecker::validateWhereClauses(ProtocolDecl *protocol,
 /// to which this type declaration conforms.
 void TypeChecker::checkInheritanceClause(Decl *decl,
                                          GenericTypeResolver *resolver) {
-  TypeResolutionOptions options;
+  TypeResolutionOptions options = None;
   DeclContext *DC;
   if (auto nominal = dyn_cast<NominalTypeDecl>(decl)) {
     DC = nominal;
-    options |= TypeResolutionFlags::GenericSignature;
-    options |= TypeResolutionFlags::InheritanceClause;
+    options = TypeResolutionOptions(TypeResolverContext::GenericSignature);
+    options.setContext(TypeResolverContext::InheritanceClause);
     options |= TypeResolutionFlags::AllowUnavailableProtocol;
   } else if (auto ext = dyn_cast<ExtensionDecl>(decl)) {
     DC = ext;
-    options |= TypeResolutionFlags::GenericSignature;
-    options |= TypeResolutionFlags::InheritanceClause;
+    options = TypeResolutionOptions(TypeResolverContext::GenericSignature);
+    options.setContext(TypeResolverContext::InheritanceClause);
     options |= TypeResolutionFlags::AllowUnavailableProtocol;
   } else if (isa<GenericTypeParamDecl>(decl)) {
     // For generic parameters, we want name lookup to look at just the
     // signature of the enclosing entity.
     DC = decl->getDeclContext();
     if (auto nominal = dyn_cast<NominalTypeDecl>(DC)) {
       DC = nominal;
-      options |= TypeResolutionFlags::GenericSignature;
+      options = TypeResolutionOptions(TypeResolverContext::GenericSignature);
```

Comment:
An implicit conversion from TypeResolverContext to TypeResolutionOptions might be nice

Reply:
Sure, that can work. FWIW – I was avoiding implicit conversions in the short term to make this refactoring work easier to reason about.

Reply:
After some thought, I'd prefer that we do not allow implicit conversion from `TypeResolverContext` to `TypeResolutionOptions` during assignment. Construction is okay though. The problem is that implicit conversion during assignment invites bugs, because everything is being reset (the base, the context, and flags).